metaseq/models/distributed_model.py

```diff
@@ -63,7 +63,10 @@ def DistributedModel(args, model, process_group, device):
         assert isinstance(model, FSDP), "expected model to already be wrapped in FSDP"
         wrapped_model = model
         if args.memory_efficient_fp16:
-            wrapped_model = wrapped_model.half()
+            if args.bf16:
```

Comment:
Nit: This seems potentially confusing to future users. Seem like we need both `memory_efficient_fp16` _and_ `bf16` turned on here, to enable the following bf16 logic.  Should we rename the `memory_efficient_fp16` flag to something more descriptive?

Reply:
+1 let's move this outside the args.mem_eff_fp16 gate

Reply:
@ngoyal2707 let's solve the backwards compatibility breakage separately - we need a lot more love to the checkpointing logic, so fixing the "breakage" would be a good starting point too!

Reply:
I've just added an assert into MemoryEfficientFP16Optimizer that asserts bf16 isn't true and tells the user not to mix the two.

Reply:
Cool, that works

Reply:
I actually am confused here lol -- why is this under `if args.memory_efficient_fp16`? cc @ngoyal2707 @suchenzang @stephenroller